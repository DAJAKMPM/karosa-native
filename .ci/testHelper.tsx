import React, { JSXElementConstructor } from "react";
import { View } from "react-native";
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from "@testing-library/react-native";
import { Formik, FormikValues } from "formik";

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };

  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || "wrapper");
  return { wrapper, ...renderApi };
};

export const renderWithFormikWrapper = <T extends FormikValues>(
  initialValues: T,
  children: React.ReactElement<any, string | JSXElementConstructor<any>>
) => {
  return renderWithWrapper(
    <Formik
      initialValues={initialValues}
      onSubmit={() => console.log("submit")}
      component={() => children}
    />
  );
};
