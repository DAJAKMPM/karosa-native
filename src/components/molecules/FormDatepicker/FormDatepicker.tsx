/**
 *
 * FormDatepicker
 * @format
 *
 */

import React, { FC, useState, useCallback } from "react";
import { useField } from "formik";
import { isNil } from "lodash";
import { useUpdateEffect } from "@app/hooks";
import Datepicker from "@app/atoms/Datepicker";

import type { PropsType } from "./types";

const FormDatepicker: FC<PropsType> = (props) => {
  const { name } = props;

  const [, meta, helpers] = useField(name);

  const [currentValue, setCurrentValue] = useState(
    isNil(meta.value) ? meta.initialValue : meta.value
  );

  useUpdateEffect(() => {
    setCurrentValue(meta.value);
  }, [meta.value]);

  const handleChange = useCallback(
    (text: string) => {
      setCurrentValue(text);
      helpers.setValue(text);
      helpers.setTouched(true);
    },
    [helpers]
  );

  return <Datepicker {...props} value={currentValue} onChange={handleChange} />;
};

export default FormDatepicker;
