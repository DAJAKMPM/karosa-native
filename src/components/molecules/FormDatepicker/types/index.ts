/**
 *
 * @format
 *
 */

import { PropsType as DatepickerType } from "@app/atoms/Datepicker/types";

type Keys = "onChange" | "value";

export type PropsType = Omit<DatepickerType, Keys> & {
  name: string;
};
