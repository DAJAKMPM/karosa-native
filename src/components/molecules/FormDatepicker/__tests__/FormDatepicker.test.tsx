import React from "react";
import { renderWithFormikWrapper } from "@app/ci/testHelper";
import FormDatepicker from "../FormDatepicker";

describe("FormDatepicker Component", () => {
  it("should match snapshot", () => {
    const { toJSON } = renderWithFormikWrapper(
      { test: null },
      <FormDatepicker name="test" />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
