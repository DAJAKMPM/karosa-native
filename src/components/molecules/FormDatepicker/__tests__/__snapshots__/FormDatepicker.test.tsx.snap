// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDatepicker Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <TextInput
    allowFontScaling={true}
    keyboardType="numeric"
    onChangeText={[Function]}
    options={
      Object {
        "format": "YYYY/MM/DD",
      }
    }
    placeholder="YYYY/MM/DD"
    rejectResponderTermination={true}
    returnKeyType="done"
    type="datetime"
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
